Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -34,6 +40,199 @@ extension BinaryInteger {
+    internal // For unit test accessibility, otherwise would be fileprivate.
+    static func decimalDigitsAndMagnitudePerWord() -> (digits: Int, magnitude: Self) {
+        // First, a fast-path that works for any type (for `Self`) which can (essentially) represent a UInt (or larger).
+        let guessDigits = Int(Double(Words.Element.bitWidth) * log10(2))
+        let guessMagnitude = pow(10, Double(guessDigits))
+
+        if let magnitudeAsSelf = Self(exactly: guessMagnitude) {
+            return (guessDigits, magnitudeAsSelf)
+        }
+
+        // Alas `Self` is smaller than UInt, so fall back to a truly generic - but slower - algorithm to find the results.  This is because BinaryIntegers - unlike e.g. FixedWidthIntegers - don't provide APIs for questions like "what is the maximum bit width of Self?".
+
+        var count = 1
+        var value: Words.Element = 1
+
+        while true {
+            var (nextValue, overflowed) = value.multipliedReportingOverflow(by: 10)
+
+            // Words.Element might be wider than the actual type, e.g. if Self is UInt8 (and UInt is not).  The magnitude is limited by the smallest of the two.
+            if !overflowed && nil == Self(exactly: nextValue) {
+                overflowed = true
+            }
+
+            if !overflowed || .zero == nextValue {
+                count += 1
+            }
+
+            if overflowed {
+                return (count - 1, Self(value))
+            }
+
+            value = nextValue
+        }
+    }
+}
+
+extension UInt {
+    /// Formats `self` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html) which is the required input form for certain ICU functions (e.g. `unum_formatDecimal`).
+    ///
+    /// This is intended to be used only by the `numericStringRepresentation` property (both the specialised form below and the generic one for all BinaryIntegers, above).  Prefer the `numericStringRepresentation` property for all other use-cases.
+    ///
+    /// - Parameter intoEndOfBuffer: The buffer to write into, which _must_ contain enough space for the result.  The formatted output is placed into the _end_ of this buffer ("right-aligned", if you will), though the output is numerically still left-to-right.  The contents of this buffer do not have to be pre-initialised.
+    /// - Returns: How many entries (UInt8s) of the buffer were used.  Note that zero is a valid return value, as nothing is written to the buffer if `self` is zero (this may be odd but it's acceptable to `numericStringRepresentation` and it simplifies the overall implementation).
+    func numericStringRepresentation(intoEndOfBuffer buffer: inout Slice<UnsafeMutableBufferPointer<UInt8>>) -> Int {
+        guard .zero != self else { return 0 } // Easier to special-case this here than deal with it below (annoying off-by-one potential errors).
```

Comment:
Nit: this case has already been handled in the caller, so the return statement is dead code in practice. Could be replaced by an assertion.

Reply:
Done.

Reply:
Actually, no, reverted.  We'd both overlooked the fact that it _can_ be called on zero because it's used for the individual words of the BinaryInteger (from the generic `numericStringRepresentation` directly, not the UInt specialisation!), any of which can actually be zero.